### PR TITLE
STORM-3312: Upgrade Guava to latest version where possible

### PR DESCRIPTION
**Pull request by srdo**
https://issues.apache.org/jira/browse/STORM-3312

### CODE REVIEW

File: examples/storm-hbase-examples/pom.xml [hunk 1]
```diff
@@ -26,6 +26,11 @@
     </parent>
 
     <artifactId>storm-hbase-examples</artifactId>
+    
+    <properties>
+        <!-- Required downgrade by hbase-client 1.4.4 -->
```
Comment (HeartSaVioR) [inline examples/storm-hbase-examples/pom.xml]: Just curious, how did you explore whether downgrade is required?

Comment (srdo) [reply examples/storm-hbase-examples/pom.xml]: It's been a while, so I might be misremembering, but I believe I used `mvn dependency:tree` and checked whether each module had any dependencies that used Guava less than version 21.0. According to the Guava README, that was the last version to remove a non-beta API (https://github.com/google/guava#important-warnings point 2). My understanding is that Guava versions are backwards compatible starting at 21.0.

The downgrade may not actually be required, but I don't want to do the testing required to figure that out. At some point someone will upgrade the direct dependency, and then we can remove the downgrade.

**Comment by srdo** [comment]
Yes, let's wait. 2.0.0 is big enough already :)

File: examples/storm-hbase-examples/pom.xml (hunk 1 above)
Comment (HeartSaVioR) [reply examples/storm-hbase-examples/pom.xml]: Thanks for explaining! I would also think this would be our best try with limited resource.

**Comment by HeartSaVioR** [comment]
@srdo Could you rebase? I think this can be handled along with STORM-3349 and STORM-3350, and still have a slot to include in 2.0.0.

**Comment by srdo** [comment]
I'll rebase soon, but I'd like to merge the Hadoop stuff first. It's likely to affect where we need to manually downgrade Guava.

**Comment by HeartSaVioR** [comment]
Yeah sure. That sounds better.

File: pom.xml [hunk 2]
```diff
@@ -266,8 +266,8 @@
         <httpclient.version>4.5.6</httpclient.version>
         <jctools.version>2.0.1</jctools.version>
         <jgrapht.version>0.9.0</jgrapht.version>
-        <guava.version>16.0.1</guava.version>
-        <auto-service.version>1.0-rc3</auto-service.version>
+        <guava.version>27.0.1-jre</guava.version>
+        <auto-service.version>1.0-rc4</auto-service.version>
```
Comment (d2r) [inline pom.xml]: Just curious, why is the version changed for `auto-service` here? Is it related to guava?

Comment (srdo) [reply pom.xml]: Yes. Rc3 uses Guava 19, while 4 uses 23.5. I think as of Guava 21, there have been no breaking changes going forward.

**Comment by srdo** [comment]
Test failure is caused by the issue fixed in https://github.com/apache/storm/pull/2943. That particular test has a habit of hitting the failing assertion in Slot for some reason.

File: pom.xml (hunk 2 above)
Comment (d2r) [reply pom.xml]: OK makes sense.

**Comment by HeartSaVioR** [comment]
Just reran the build which failed. Will +1 again and merge once the build passes.

**Comment by srdo** [comment]
@HeartSaVioR Thanks for reviewing. I'll rebase this and other PRs now that the Slot fix is merged. Otherwise I suspect the test will keep failing.

**Comment by HeartSaVioR** [comment]
@srdo Ah yes right. Travis CI doesn't seem to try to merge the change with latest master. Maybe better to have that but that would not be easy with free CI options. Will see whether tests on PRs are passing.

**Comment by srdo** [comment]
I was wrong, it sounds like Travis does merge with master before building https://docs.travis-ci.com/user/pull-requests/#how-pull-requests-are-built. Still have to trigger the builds manually after changes on master though.

**Comment by HeartSaVioR** [comment]
Good to hear if Travis CI can deal with it (TBH a bit confused here - not sure why I thought it didn't): that means we could rerun the build again just before merging to ensure it also works with latest master.

**Comment by srdo** [comment]
Yes. Might be good if PRs are ready to merge for a while, but most of the time I think PRs are only made ready to merge shortly before they actually get merged, so in most cases there shouldn't be a need to do it.

**Comment by HeartSaVioR** [comment]
Sure, not intended to bring more steps/process here. We just found safer way to merge. :)

**Comment by srdo** [comment]
:thumbsup:

**Pull request merged**